rust-runtime/aws-smithy-http-server/src/rejection.rs

```diff
@@ -0,0 +1,152 @@
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Rejection response types.
+use axum::response::IntoResponse;
+
+define_rejection! {
+    #[status = INTERNAL_SERVER_ERROR]
+    #[body = "Cannot have two request body extractors for a single request"]
+    /// Rejection type used if you try and extract the request body more than
+    /// once.
+    pub struct BodyAlreadyExtracted;
+}
+
+define_rejection! {
+    #[status = INTERNAL_SERVER_ERROR]
+    #[body = "Headers taken by other extractor"]
+    /// Rejection used if the headers has been taken by another extractor.
+    pub struct HeadersAlreadyExtracted;
+}
+
+define_rejection! {
+    // Should this be INTERNAL_SERVER_ERROR ?
+    #[status = BAD_REQUEST]
+    #[body = "Request deserialize failed"]
+    /// Rejection type used if the request deserialization encountered errors.
+    pub struct Deserialize(Error);
+}
+
+define_rejection! {
+    #[status = INTERNAL_SERVER_ERROR]
+    #[body = "Response serialize failed"]
+    /// Rejection type used if the response serialization encountered errors.
+    pub struct Serialize(Error);
+}
+
+define_rejection! {
+    #[status = BAD_REQUEST]
+    #[body = "Request body do not contain valid UTF-8"]
+    /// Rejection type used when buffering the request into a [`String`] if the
+    /// body doesn't contain valid UTF-8.
+    pub struct InvalidUtf8(Error);
+}
+
+define_rejection! {
+    // TODO: we probably want to be more specific as the http::Error enum has many variants
+    #[status = INTERNAL_SERVER_ERROR]
+    #[body = "Error handling HTTP request"]
+    /// Rejection type used when there is an error handling the HTTP request
+    pub struct Http(Error);
+}
+
+define_rejection! {
+    // TODO: we probably want to be more specific as the header parsing can have many variants
+    #[status = BAD_REQUEST]
+    #[body = "Error parsing headers"]
+    /// Rejection used if the any of the header parsing fails.
+    pub struct HeadersParse(Error);
+}
+
+define_rejection! {
+    #[status = BAD_REQUEST]
+    #[body = "Expected `Content-Type: application/json`"]
+    /// Rejection type used if the JSON `Content-Type` header is missing.
+    pub struct MissingJsonContentType;
+}
+
+define_rejection! {
+    #[status = BAD_REQUEST]
+    #[body = "Failed to parse request MIME type"]
+    /// Rejection type used if the MIME type parsing failed..
+    pub struct MimeParsingFailed;
+}
+
+composite_rejection! {
+    /// Rejection used for content-type errors such has missing content-type
+    /// header, MIME parse issues, etc..
+    pub enum ContentTypeRejection {
+        MissingJsonContentType,
+        MimeParsingFailed,
+    }
+}
+
+composite_rejection! {
+    /// General rejection used as Rejection type in the Axum FromRequest / IntoResponse traits.
+    ///
+    /// Contains one variant for each way the Smithy Rust Server can fail. It doesn't account for
+    /// customer defined operation errors.
```

Comment:
```suggestion
    /// Contains one variant for each way the Smithy Rust server can fail.
```

It gives the impression business logic errors are Smithy Rust server errors, but they're not included here. I'd just leave like this.

Reply:
Ok, I see that it can be confusing..